mkosi/cage.py

```diff
@@ -0,0 +1,664 @@
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+"""
+This is a standalone implementation of sandboxing which is used by mkosi. Note that this is invoked many times while
+building the image and as a result, the performance of this script has a substantial impact on the performance of mkosi
+itself. To keep the runtime of this script to a minimum, please don't import any extra modules if it can be avoided.
+"""
+
+import ctypes
+import encodings.ascii  # noqa: F401 (loaded lazily by open() which happens too late)
+import os
+import sys
+import warnings  # noqa: F401 (loaded lazily by os.execvp() which happens too late)
+
+AT_EMPTY_PATH = 0x1000
+AT_FDCWD = -100
+AT_NO_AUTOMOUNT = 0x800
+AT_RECURSIVE = 0x8000
+AT_SYMLINK_NOFOLLOW = 0x100
+BTRFS_SUPER_MAGIC = 0x9123683E
+CAP_NET_ADMIN = 12
+CAP_SYS_ADMIN = 21
+CLONE_NEWIPC = 0x08000000
+CLONE_NEWNET = 0x40000000
+CLONE_NEWNS = 0x00020000
+CLONE_NEWUSER = 0x10000000
+ENOENT = 2
+LINUX_CAPABILITY_U32S_3 = 2
+LINUX_CAPABILITY_VERSION_3 = 0x20080522
+MNT_DETACH = 2
+MOUNT_ATTR_RDONLY = 0x00000001
+MOUNT_ATTR_NOSUID = 0x00000002
+MOUNT_ATTR_NODEV = 0x00000004
+MOUNT_ATTR_NOEXEC = 0x00000008
+MOUNT_ATTR_SIZE_VER0 = 32
+MOVE_MOUNT_F_EMPTY_PATH = 0x00000004
+MS_BIND = 4096
+MS_MOVE = 8192
+MS_REC = 16384
+MS_SHARED = 1 << 20
+MS_SLAVE = 1 << 19
+NR_mount_setattr = 442
+NR_move_mount = 429
+NR_open_tree = 428
+OPEN_TREE_CLOEXEC = os.O_CLOEXEC
+OPEN_TREE_CLONE = 1
+PR_CAP_AMBIENT = 47
+PR_CAP_AMBIENT_RAISE = 2
+SCMP_ALLOW = 0x7FFF0000
+SCMP_ERRNO = 0x00050000
+
+class mount_attr(ctypes.Structure):
+    _fields_ = [
+        ("attr_set", ctypes.c_uint64),
+        ("attr_clr", ctypes.c_uint64),
+        ("propagation", ctypes.c_uint64),
+        ("userns_fd", ctypes.c_uint64),
+    ]
+
+class cap_user_header_t(ctypes.Structure):
+    _fields_ = [
+        ("version", ctypes.c_uint32),
+        ("pid", ctypes.c_int),
+    ]
+
+class cap_user_data_t(ctypes.Structure):
+    _fields_ = [
+        ("effective", ctypes.c_uint32),
+        ("permitted", ctypes.c_uint32),
+        ("inheritable", ctypes.c_uint32),
+    ]
+
+libc = ctypes.CDLL(None, use_errno=True)
+
+libc.syscall.restype = ctypes.c_long
+libc.unshare.argtypes = (ctypes.c_int,)
+libc.statfs.argtypes = (ctypes.c_char_p, ctypes.c_void_p)
+libc.eventfd.argtypes = (ctypes.c_int, ctypes.c_int)
+libc.mount.argtypes = (ctypes.c_char_p, ctypes.c_char_p, ctypes.c_char_p, ctypes.c_ulong, ctypes.c_char_p)
+libc.pivot_root.argtypes = (ctypes.c_char_p, ctypes.c_char_p)
+libc.umount2.argtypes = (ctypes.c_char_p, ctypes.c_int)
+libc.capget.argtypes = (ctypes.c_void_p, ctypes.c_void_p)
+libc.capset.argtypes = (ctypes.c_void_p, ctypes.c_void_p)
+
+
+def oserror(arg: str = "") -> None:
+    if arg:
+        raise OSError(ctypes.get_errno(), os.strerror(ctypes.get_errno()), arg)
+    else:
+        raise OSError(ctypes.get_errno(), os.strerror(ctypes.get_errno()))
```

Comment:
Let's stay with the variable name that `OSError` uses and since the additional arguments default to `None`, we can shorten this a bit.
```suggestion
def oserror(filename: str = "") -> None:
    raise OSError(ctypes.get_errno(), os.strerror(ctypes.get_errno()), filename or None)    
```